[ClangImporter] Don't use local scope decls to find a function's home

jrose-apple opened the pull request:
In C, a function can be declared at local scope:

```c
void aFunctionInBase(void) {
  void theFunctionInQuestion(int);
}
```

and then again in a different header at top-level scope:

```c
void theFunctionInQuestion(int);
```

If the first one appears first, it becomes what Clang considers the "canonical" declaration, which (up until now) Swift has been using to decide what module to import a function into. (Since a C function can be redeclared arbitrarily many times, we have to pick one.) This is important for diagnostics and anything else that might ask "where did this Swift declaration come from". Instead of the very first redeclaration, use the first non-local one to determine the "home" module.

(The standard library wants a guarantee that forward declarations they put in SwiftShims won't interfere with declarations found elsewhere. I don't think Clang can *ever* provide that, so if there's ever a mismatch between the standard library's forward declarations and the "real" declarations the standard library's might win out at the LLVM level—say, in terms of attributes. But this at least removes a place where that could be visible to users even when it isn't otherwise a problem.)

jrose-apple commented:
cc @stephentyrone, @milseman, @Catfish-Man 

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

stephentyrone commented:
Can we build a simple test that barfs if any of these symbols gets attributed to the shims module?

varungandhi-apple commented on lib/ClangImporter/ClangAdapter.h:
```diff
@@ -60,6 +60,16 @@ struct PlatformAvailability;
+/// Use this similarly to clang::Decl::getCanonicalDecl when the containing
+/// module matters.
+const clang::Decl *
+getFirstNonLocalDecl(const clang::Decl *D);
```
Does "NonLocal" == "top-level"? Or is there a case where there's a difference between the two? In case, they're the same, it might be more understandable to rename the function to `getFirstTopLevelDecl` (at first glance, I couldn't tell what "nonlocal" mean).

I think it would be helpful to provide a little bit of information from the commit message here in the doc comment. For example, the reader might not know that C even allows declarations inside functions, and that clang's "canonical" location considers those, which is why it is not appropriate here.

varungandhi-apple commented on lib/ClangImporter/ClangAdapter.h:
```diff
@@ -60,6 +60,16 @@ struct PlatformAvailability;
+/// Returns the first redeclaration of \p D outside of a function.
+///
+/// If there are no such redeclarations, returns null.
+/// If \p D is not redeclarable, just returns \p D itself.
```
What does "is not redeclarable" mean -- which declarations are redeclarable and which ones are not? Is this a well-known thing? :confused:

jrose-apple replied:
It's Clang terminology, but clearly not clear enough.

varungandhi-apple commented on lib/ClangImporter/ClangAdapter.h:
```diff
@@ -60,6 +60,16 @@ struct PlatformAvailability;
+/// If \p D is not redeclarable, just returns \p D itself.
+///
+/// Use this similarly to clang::Decl::getCanonicalDecl when the containing
+/// module matters.
```
In what situations does the containing module not matter? Is this supposed to be clear from the usage context?

jrose-apple replied:
:-/ I don't have an answer to this one. I looked at each of the call sites for `getCanonicalDecl` and determined that only one made sense to change because all the others weren't going to ask the question "which module are you from", but still.

jrose-apple commented:
> Can we build a simple test that barfs if any of these symbols gets attributed to the shims module?

No. The symbols *will* get attributed to the shims module from Clang's perspective, barring modules bugs—it will be the first declaration seen in every single compilation except those with bridging headers. I *think* it's the most recent declaration that matters in most cases anyway, but modules throw that out the window a bit.

jrose-apple committed:
[ClangImporter] Don't use local scope decls to find a function's home

In C, a function can be declared at local scope:

    void aFunctionInBase(void) {
      void theFunctionInQuestion(int);
    }

and then again in a different header at top-level scope:

    void theFunctionInQuestion(int);

If the first one appears first, it becomes what Clang considers the
"canonical" declaration, which (up until now) Swift has been using to
decide what module to import a function into. (Since a C function can
be redeclared arbitrarily many times, we have to pick one.) This is
important for diagnostics and anything else that might ask "where did
this Swift declaration come from". Instead of the very first
redeclaration, use the first non-local one to determine the "home"
module.

(The standard library wants a guarantee that forward declarations they
put in SwiftShims won't interfere with declarations found elsewhere. I
don't think Clang can /ever/ provide that, so if there's ever a
mismatch between the standard library's forward declarations and the
"real" declarations the standard library's might win out at the LLVM
level---say, in terms of attributes. But this at least removes a place
where that could be visible to users even when it isn't otherwise a
problem.)

DougGregor reviewed (approved):
Huh, I missed this before. Thank you!

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please test Linux

varungandhi-apple commented:
@swift-ci please test
